src/main/java/net/neoforged/neoforge/common/util/GlobalVec3.java

```diff
@@ -5,51 +5,295 @@
+        this(dimension, new WeakReference<>(null), vec3);
     }
 
     public GlobalVec3(Entity entity) {
         this(entity.level(), entity.position());
     }
 
     public GlobalVec3(Level level, BlockPos blockpos) {
-        this(level, blockpos.getX(), blockpos.getY(), blockpos.getZ());
+        this(level.dimension(), new WeakReference<>(level), blockpos.getX(), blockpos.getY(), blockpos.getZ());
+    }
+
+    public GlobalVec3(ResourceKey<Level> dimension, BlockPos blockpos) {
+        this(dimension, new WeakReference<>(null), blockpos.getX(), blockpos.getY(), blockpos.getZ());
+    }
+
+    public GlobalVec3(GlobalPos globalpos) {
+        this(globalpos.dimension(), globalpos.pos());
     }
 
+    // setters
+
     public GlobalVec3 withX(double x) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(dimension, level, x, y, z);
     }
 
     public GlobalVec3 withY(double y) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(dimension, level, x, y, z);
     }
 
     public GlobalVec3 withZ(double z) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(dimension, level, x, y, z);
     }
 
     public GlobalVec3 withLevel(Level level) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(level.dimension(), new WeakReference<>(level), x, y, z);
+    }
+
+    public GlobalVec3 withDimension(ResourceKey<Level> dimension) {
+        return new GlobalVec3(dimension, new WeakReference<>(null), x, y, z);
+    }
+
+    // getters
+
+    public ResourceKey<Level> getDimension() {
+        return dimension;
+    }
+
+    /**
+     * Returns the level of this global position if it is available.
+     * <p>
+     * Use {@link #getLevel(MinecraftServer)} or {@link #getLevel(Level)} if possible.
+     * 
+     * @return The level.
+     */
+    public @Nullable Level getLevel() {
+        return level.get();
+    }
+
+    public Level getLevel(Level anotherLevel) {
+        return getLevel(anotherLevel.getServer());
+    }
+
+    public Level getLevel(MinecraftServer server) {
+        final Level level2 = level.get();
+        if (level2 == null) {
+            return server.getLevel(dimension);
+        }
+        return level2;
+    }
+
+    // converters
+
+    public Vec3 toVec3() {
+        return super.add(0, 0, 0);
+    }
+
+    public BlockPos toBlockPos() {
+        return BlockPos.containing(this);
+    }
+
+    public GlobalPos toGlobalPos() {
+        return new GlobalPos(dimension, toBlockPos());
+    }
+
+    public GlobalVec3 copy() {
+        return new GlobalVec3(dimension, level, x, y, z);
+    }
+
+    // method overrides
+
+    protected void sameDim(Vec3 other) {
+        if (other instanceof GlobalVec3 other3 && other3.dimension != dimension) {
+            throw new IllegalStateException("Cannot compute with GlobalPoses that are in different dimensions");
```

Comment:
Should we add the dimension keys of the two global vectors in the exception message, to ease debugging?